### ISSUE TITLE
Discussion on Angular experiment

### ISSUE THREAD

**Issue by christianalfoni** [christianalfoni/flux-angular#28]
Hi @sheerun, @sebakerckhof, @mlegenhausen, @jolafrite, @SuperheroicCoding, 

I have pre-published an article I was hoping you would take the time to read. Then we can discuss if you see any advantages of bringing any of those concepts into a new version of flux-angular.

Flux is evolving and I think it is a good idea to keep flux-angular in the loop, though that means API changes.

Anyways, article is over here: [The great Angular component experiment](http://christianalfoni.github.io/javascript/2015/02/22/the-great-angular-component-experiment.html).
 
Thanks :-)

**Comment by sebakerckhof** [comment]
Do we discuss it here or do I post in the comment section on the article?

**Comment by jolafrite** [comment]
Hi @christianalfoni 

I had a look at your new version of flux-angular.
I need to spend some time to play around and see deeper but from the first look it looks interesting.

Thanks

By the way there is some typos on your blog article.

**Comment by christianalfoni** [comment]
@sebakerckhof Would be great to have the discussion here on GitHub as it is flux-angular specific :-)

@jolafrite Jup, cool. Please look at the article as an experiment, not a specific suggestion for the next version. More use the ideas from the article, if you guys feel it makes sense

I usually read through an article 20 times before publishing, even then it takes about a week before typos are gone, hehe. Done a few updates already, so please let me know if you find any. Quick fix :-)

**Comment by mlegenhausen** [comment]
@christianalfoni thanks for this article, I read all your previous articles about flux before and they let me think a lot about it. I had the same ideas before about the components architecture. I read some articles about component design in angular before. I think this is the way to go in the future (or as soon as possible).

Can you provide your implementation for the `components` and `actions` methods. What are you thoughts about this components implementation:

```js
app.directive ('component', function () {
  return {
    controllerAs: 'component',
    scope: scope,
    template: template,
    controller: controller
  };
});
```

shortcut for

```js
app.component ('component', template, scope, controller);
```

proposed from egghead.

I have some issues with the way you think about actions. In your todo example with one store everything is fine, but when you application grows you want to have more stores that store a specific typ of state of your application and an action is a great way to provide all information that is needed for the state change to all of this different stores. How do you scale your actions for bigger applications?

**Comment by sebakerckhof** [comment]
Okay, here are my 2¢:
Flux, and the derivative you talk about in the article are high-level concepts that are mainly about having a certain mindset as a developer. You try to enforce this mindset by monkeypatching the angular object with component / store / actions, and for this I don't see much use. You could just as well create small, focussed and isolated components using the normal directive API with the right mindset. Just like you can probably create very bulky components using your component API with the wrong mindset.

And to a large extend you could just create the store as an object on the rootscope and make a service that provides mutation functions to mutate the object on the rootscope. So then you would have directives / rootscope object / mutation service . You'd lose immutability, but I still don't see the advantage of immutability in Angular.

**Comment by mlegenhausen** [comment]
You don't have to think about what brings immutability for angular. Instead of what brings it in generell.

**Comment by sebakerckhof** [comment]
And a little typo in this sentence: "means that neither your or Angular can change objects"

**Comment by christianalfoni** [comment]
@mlegenhausen Thanks for sharing the suggestion from egghead.io. I can not see how you could make it cleaner without wrapping it, like your next code example. Since I was taking this a step further, to take inspiration from React JS, a component is actually the $scope itself.

This is how it looks:
```javascript
moduleInstance.component = function (name, spec) {

    moduleInstance.directive(name, ['$parse', '$injector', function ($parse, $injector) {

      var data = $injector.invoke(spec);

      return {
        restrict: 'E',
        transclude: true,
        replace: true,
        compile: function () {
          return {
            pre: function (scope, element, attrs) {

              scope.getDOMNode = function () {
                return null;
              };
              scope.props = {};

              // Add props to the scope
              Object.keys(attrs).forEach(function (key) {
                if (key[0] !== '$' && key.substr(0, 2) !== 'ng') {
                  scope.props[key] = $parse(attrs[key])(scope) || attrs[key];
                }
              });

              // 
              Object.keys(data).forEach(function (key) {
                if (key !== 'componentWillUnmount' && key !== 'componentWillMount' && key !== 'componentDidMount' && key !== 'render' ) {
                  scope[key] = data[key];
                }
              });

              if (data.componentWillMount) {
                data.componentWillMount.call(scope);
              }

              scope.$on('$destroy', function () {
                if (data.componentWillUnmount) {
                  data.componentWillUnmount.call(scope);
                }
              });

            },
            post: function (scope, element) {

              scope.getDOMNode = function () {
                return element;
              };

              if (data.componentDidMount) {
                setTimeout(function () {
                  data.componentDidMount.call(scope);
                }, 0);
              }
              
            }
          }
        },
        template: typeof data.render === 'function' ? data.render : null,
        templateUrl: typeof data.render === 'string' ? data.render : null
      }
    }]);

    return this;

  };
```

It overrides a lot of Angulars default behavior, so I totally agree with you guys, also @sebakerckhof, that this is taking it too far. It was just fun to see how it would look if you think of it as a React JS component :-)

What I think we can take out of this though is the egghead.io suggestion. Document it on the repo. Give an example of how it is used. Could of course also include a `.component` monkeypatch, but I think the signature derives too much from the general signature (a function), and you would probably need other stuff like compile etc. to handle pre-compile post-compile stuff. Then you would have two different signatures for components. So yeah, summarized. I can document the egghead.io suggestion in the readme and give examples of how to use it :-) Thanks for that!

## Store
The store concept in the experiment attaches a domain name to the root scope. The domain name (todos in the example) is a store by itself. You could continue adding more stores:

```javascript
angular.module('app', ['flux'])
  .store('todos', function () {...})
  .store('notifications', function () {...})
  .store('users', function () {...});
```

The great thing about this is that we have an API for attaching stuff to $rootScope. And the good thing about attaching stuff to the $rootScope is that it is the same principle as injecting the stores into the application. You do not depend on them in the components at all. That makes components even easier to test and less dependent.

That said, this is a concept from React JS and FLUX that allows for isomorphic apps. That is not really possible with Angular, as you can not render Angular on the server side. That said, the concept of having clean components and inject state "at the top" I think is a really good concept.

## Actions
The actions in the experiment were quite simple, but they are really no different than "Action Creators" you would create for the current version of flux-angular to handle async stuff. You can have complex updates of the store. The really good thing with the experiment is that you have one entry point for all your state. The current flux-angular implementation requires you to use the `waitFor` method to make stores wait for each other. That can become quite complex. There is also the issue where a handler is basically just mutating the state in the store. It is a bit verbose to create handlers for just changing `isLoading = true`, `items.push` etc. Putting the line you always need to notify that a change is done makes it even more verbose. Not saying that it is bad at all, just saying what I think can be an advantage taking inspiration from the experiment :-)

## Immutability
I think of immutability much like type checking. You do not need any of them, as long as you know what you are doing, and you do not get to a certain level of complexity. There is a reason why Facebook spends so much time on immutability and type checking, they have huge projects, with huge teams... but most of us do not work that way. 

The problem with using FLUX in Angular is its two-way-data-binding. It has it advantages, but not when thinking application state flow. It is a lot easier thinking of your application state as flowing down into components and can only be changed by an action. As soon as you start "ng-model"-ing up your application state changes can happen everywhere. This is also not a problem until you reach a certain level of complexity, but I think the level is quite low, at least in my experience.

What immutability lets us do is block this behavior. We still allow it between the component $scope properties and the template, but we do not allow it between a template/$scope and the stores. You should be in full control of when a store changes, and that is through actions.

The **immutable-store** used in the experiment is a lot faster than the current flux-angular store, because it does not use cloning to create an immutable environment. We could replace the current store with **immutable-store** though. Get some more performance and a less verbose API.

## A suggestion
Okay, so based on the discussion so far I will give a suggestion on API where we use egghead.io components and the new store. Just to start somewhere. Please point out if you are perfectly happy with the API as it works now too :-)

```javascript
angular.module('app', [])

   // Egghead definition
  .directive('myComponent', {
    return {
      controllerAs: 'component',
      scope: {callback: '&', title: '=', message: '@'},
      templateUrl: 'myComponent.html',
      controller: function ($scope) {
        $scope.sayHi = function () {
          alert('hi ' + $scope.user.username);
        };
      }
    };
  })

  // Attaches state to $rootScope with store name ("user")
  .store('user', function () {
    return {
      username: 'something',
      age: 30
    };
  })

  // Actions are just factories 
  .factory('actions', function (flux) {
    return {
      changeAge: function (newAge) {
        
        // Any change to store objects and arrays returns
        // a change that flux needs to know about. That triggers
        // the actual update on the $rootScope
        var stores = flux.get();
        var change = stores.user.set('age', newAge);
        flux.set(change);

        // You can also emit events as normal of course
        flux.emit('my-event');
      }
    };
  })
  .run(function (flux) {

    // You can easily attach state with the initial response,
    // making your app render faster
    flux.set(window.bootstrappedStoreData);

    // Or put state in local storage, even faster
    window.addEventListener('beforeleave', function () {
      localStorage.store = JSON.stringify(flux.get());
    });
  });
```

Okay, so what this first draft gives:

1. An injected store concept so you do no have to listen to stores in controllers/directives. They are part of the scope by store name. This makes components more pure, less verbose and easier to test

2. The store is a lot faster as it does not use cloning, but still uses immutability to give you a one way flow of state. You have to embrace the concept that a mutation returns a change that needs to be set

3. A component concept is taking it a bit too far, but documentation on how to ideally do it and still give flexibility is added to the documentation

**Comment by christianalfoni** [comment]
btw, @sebakerckhof , thanks for typo. Fixed it now :-)

**Comment by christianalfoni** [comment]
You know what, I have been thinking a bit more and the one thing I want to solve is to allow for switching between cloning and immutability. The [immutable-store](https://github.com/christianalfoni/immutable-store) is very small and it will make flux-angular a lot faster. The API stays exactly the same, the only difference is something like:

```javscript
angular.module('app', ['flux'])
.store('AppStore', function (flux) {

  // Create immutable state tree
  var state = flux.immutable({
    comments: {
      list: []
    }
  });

  return {
    handlers: {
      'addComment': 'addComment'
    },
    addComment: function (comment) {
      state = state.comments.list.push(comment);
      this.emitChange();
    },
    exports: {
      get comments() {
        return state.comments.list;
      }
    }
  };
})
.run(function (flux) {
  flux.configure({
    cloning: false
  });
});
```
I think I will go ahead and put this in, as it will be an addition to the current API. Let you try it out :-)

There are two performance benefits to this:

1. Now the export methods does not clone at all, it will be normal getters

2. Angular will be able to recognize an unchanged value. So returning the list in the example above, when it is unchanged, will not make Angular render the list again

Let me know if you have any comments on this.

**Comment by SuperheroicCoding** [comment]
Nice one! Thanks for your work and time!
I see the benefits of immutability and copying data only when actually changing.
I know nothing about the immutability implementation so one question:

Can I still work with the immutable object that I get from the store or do I have to set cloning : true? 
Something like: 
```javascript
var person = store.getPerson(); //getting a person from an immutable store state.
person.name = 'New Name'; //changing the persons name (local)
//... changing any other attributes... 
flux.dispatch('updatePerson',person); // dispatching to store... //update global
```
Is guess this is not possible anymore, because the person is immutable.

I prefer to change the persons name outside of the store and then dispatch the changed person object to the store. Of cause only after dispatching the action the change is visible for all depending components. 
Did you consider to use a provider to configure flux on bootstrap and
what do you think about setting this behavior for each store individuell on creation (optional)?

**Comment by christianalfoni** [comment]
Hi @SuperheroicCoding,

Thanks a lot for the feedback :-)

Yes, you are correct that you are not able to change the name. But if you think of this a bit differently:

```javascript
var person = store.getPerson(); //getting a person from an immutable store state.
flux.dispatch('updatePerson', person, {
  name: 'new name'
}); // dispatching to store... //update global
```
And your handler would then merge the changes instead. **immutable-store** does not currently have a merge method, but I could add that. So in your handler you would do:

```javascript
handlers: {
  updatePerson: 'updatePerson'
},
updatePerson: function (person, change) {
  state = person.merge(change);
  this.emitChange();
}
```
How would you feel about that? :-)

Ah, yes, a provider would of course be the correct way of implementing the configuration. It is noted!

**Comment by SuperheroicCoding** [comment]
Yes, that feels fluxy! :)
Did you test if angular ng-repeat or forms and ng-model can handle this immutable objects?

**Comment by christianalfoni** [comment]
Hehe, good, I agree :-)

You can not use ng-model with these objects/arrays, as you do not want those objects/arrays to be updated by two-way-databinding. This is not really a problem when updating one value, e.g. a silly example:

```javascript
angular.module('app', ['flux'])
  .controller('MyCtrl', function (MyStore, $scope) {
    var firstTodo = MyStore.notifications[0];
    $scope.title = firstTodo.title;
    $scope.submit = function () {
      flux.dispatch('updateTodoTitle', firstTodo, $scope.title);
    };
  });
```

But I can see it becomes tedious if you have like a person and want to put those details in a form to do updates. It is very typical for immutable APIs to have a `.toJS()` method. So to make an elegant solution for this i suggest:

 ```javascript
angular.module('app', ['flux'])
  .controller('MyCtrl', function (MyStore, $scope) {
    var person = MyStore.currentPerson;
    $scope.person = person.toJS(); // Allows you to bind, since it is a regular object
    $scope.submit = function () {
      flux.dispatch('updatePerson', person, $scope.person);
    }
  });
```
So in summary. **immutable-store** needs a `.merge()` and `.toJS()` method :-) What do you think?

**When it comes to ng-repeat** it is not problem as long as you `track by` something. Either $index or id on object. Angular will not be able to insert those hashes, when you do not track, which is actually a pain when you mostly want to put those objects on the server.

**Comment by jolafrite** [comment]
Thanks for this new version
It will be smooth to migrate to an immutable version of stores (compare to your previous proposition)
By the way, do you plan to add back component ?

**Comment by christianalfoni** [comment]
Hi @jolafrite ,

I am not quite sure. Some people think it is overkill, but others sees it as a benefit. Its also a question of what should be underneath the implementation. In the experiment I am taking it too far I think, but it would be interesting to see how the egghead.io suggestion would work under the implementation of a component. I am going to check that out and give a suggestion here :-)

**Comment by christianalfoni** [comment]
Okay guys, I have been researching a bit. This is what I have come up with. It is a wrapper around the egghead.io recommendation with a small tweak. This is the syntax:

```javascript
    .component('myComponent', function(MyStore, flux) {
          
          this.title = '';
          this.items = MyStore.items;

          this.addItem = function () {
            flux.dispatch('addItem', this.title);
            this.title = '';
          };

          this.listenTo(MyStore, function () {
            this.items = MyStore.items;
          });
      
    });
```
It uses the `bindToController` property of Angular 1.3. So in the template you point to `<button ng-click="myComponent.addItem()">Add item</button>`. By pointing to the component/controller by name it is easier to create nested components that depend on each other. Like a tab could easily point to a tabs parent component etc. 

To bind attributes to the component you can pass an object:

```javascript
    .component('myComponent', {
     someAttr: '@',
     someFunc: '&'
    }, function(MyStore, flux) {
          
          this.title = '';
          this.items = MyStore.items;

          this.addItem = function () {
            flux.dispatch('addItem', this.title);
            this.title = '';
          };

          this.listenTo(MyStore, function () {
            this.items = MyStore.items;
          });
      
    });
```

I have also given the prototype of the controller a `listenTo` method. As this has nothing to do with scope so I chose to remove the $ prefix. Gives it a cleaner look too.

**So where is the HTML?**
I have created a convention. By default it will try to load up `myComponent.html`, since that is the name of the component, but you can define a specific path to your templates using the provider:

```javascript
    .config(function (fluxProvider) {
      fluxProvider.componentTemplatePath('templates/components/');
    });
```

This will of course look for `templates/myComponent.html`.

What you think?

**Comment by christianalfoni** [comment]
Ops, maybe I should reference you guys here :-)

@sebakerckhof , @jolafrite , @mlegenhausen , @sheerun , @SuperheroicCoding

**Comment by jrust** [comment]
I have a few thoughts:
1. Immutability is really nice for the performance benefits, but definitely should be opt-in as it introduces gotchas.  And it's not clear at what point you would see a real performance hit for deep cloning.
2. I like the idea of being able to dehydrate and rehydrate state, but am not sure anything needs to be changed.  We're actually already doing this without changing anything in flux-angular right now by using Yahoo's dispatchr stuff that's already baked in with `flux.dispatcher.dehydrate` and `flux.dispatcher.rehydrate`
3. I'm not a huge fan of the component idea because it feels too much like trying to emulate react instead of implementing flux.  I'm also reticent to add another angular API modification when the angular API is already so massive and complex (and critiqued for that).  The `store` addition seems great because it's really quite hard to implement in native angular, but the `component` addition seems like only a thin wrapper around a certain type of directive and so far I haven't experienced the view layer as a pain point.

**Comment by christianalfoni** [comment]
Hi @jrust and thanks for contributing to the discussion!

1. I totally agree

2. Good point, this should be documented (Will see to that)

3. Hm, yeah, its a good point and I was actually happy with the egghead.io idea. The thing is that I was hoping to achieve listening where you use `bindToController`. But it really is not problem just using $scope, which you do have access to. 

Okay, I will leave this discussion for now... and update with new version of immutable store ;-)

**Comment by jrust** [comment]
I guess I haven't found the need for `bindToController` because I am trying to pass all the properties from stores into the directives so that they don't even have to know about the stores.  I do have thin controllers that have knowledge of the stores and a `listenTo`.  By the way, could you post a link to the egghead.io article referenced above?  Thanks.

**Comment by christianalfoni** [comment]
@mlegenhausen , where did you find that egghead.io proposed component? Cant find it :-) Would like to point to it from the readme

**Comment by mlegenhausen** [comment]
I referenced it from [another reference](http://blog.ninja-squad.com/2014/12/15/what-is-coming-in-angularjs-1.4/). Don't know the original proposal ;)

**Comment by christianalfoni** [comment]
Cool @mlegenhausen , thanks, will use that :-)
